### PR TITLE
Add sounds for falling and attached nodes

### PR DESCRIPTION
**Pull request by DS-Minetest**
As requested in #6355.

The sound specified in `node_definition.sounds.fall` is used.
When a falling node is spawned at a position or an attached node drops, the sound is played at this position.

I've tested this with `default:sand` and `default:torch` by adding in `minetest_game/mods/default/functions.lua` to `default.node_sound_defaults` the lines:
```lua
	table.fall = table.fall or
			{name = "default_break_glass", gain = 1.0}
```

### CODE REVIEW

**Comment by paramat** [comment]
A sound when a node first starts to fall seems low-priority and unnecessary, that's a suggestion in the issue i disagree with. We already have a sound when a falling node hits the ground.
It is realistic but if we work at this level of sound-detail there are many more types of sound that will have to be added. We have to pick a minimal number of primary sounds.

Adding another sound type to nodes will greatly increase sound creation work (which is the most time consuming work in MTG) for no good reason. The fall sound should use the existing 'dig' or 'dug' sound, those are are suitable.

An attached node starts to fall when another node is dug, so the sound will be covered up by the 'dug' sound of the parent node. Realistically the sound created when an attached node starts to fall is the sound of the parent node being removed, there's no need for another sound.
:-1:

**Comment by Wuzzy2** [comment]
> low-priority and unnecessary

Invalid. This is a PR with actual code, not just a feature request.

> Adding another sound type to nodes will greatly increase sound creation work

@paramat, what the hell are you talking about?!
None of these sounds are a requirement. As I understand it, this PR opens up the **possibility** that game authors **can** (read: optionally) add a sound in the event a node starts to drop. Nobody will be forced (not even MTG) that these sounds are added. If these sounds are not used in MTG, there will be simply no sound (which is a totally reasonable default). Thus, this change is 100% backwards-compatible.
Why is it bad that mods can add more node sounds if they want?

> The fall sound should use the existing 'dig' or 'dug' sound, those are are suitable.

But that is missing the point and is limiting artistic freedom. For some games, yes, the dig/dug sound might be suitable, but clearly not all of them. It's also an ugly surprise to modders to learn the hard way that the dig sound is *also* sound for falling nodes.
It should not be the job of the engine to dictate what kind of artwork is “allowed”. ;-)

If you don't want such sounds in MTG, just don't add them in MTG. Not having a fall sound is a perfectly valid artistic choice. So is *having* a fall sound. Just because MTG does not need them does not mean that NO game ever would want them.

A good engine should allow for more uses. This ideology of “if it doesn't benefit MTG, it is not needed” is really damaging for Minetest.

> Realistically the sound created when an attached node starts to fall is the sound of the parent node being removed, there's no need for another sound.

Again, this is something we should artists to decide. Maybe another sound is not needed for game A, but maybe it will work out nicely for game B. We simply cannot decide as an engine. Actual game creators probably know much better.

Let me put it another way: “Artistic freedom”. Minetest should allow for more artistic freedom, not limit it (and even less so intentionally!).

> It is realistic but if we work at this level of sound-detail there are many more types of sound that will have to be added. We have to pick a minimal number of primary sounds.

What kind of logic is that? This change is literally 9 lines of code long. Also, why does it have to be minimalistic in the first place? As long as all sounds we allow are optional or have reasonable fallbacks, I don't see a problem in extending *per se*.

File: builtin/game/falling.lua [hunk 1]
```diff
@@ -170,6 +175,9 @@ local function drop_attached_node(p)
 		drops = drop_stacks
 		def.preserve_metadata(pos_copy, node_copy, oldmeta, drops)
 	end
+	if def and def.sounds and def.sounds.fall and def.sounds.fall.name then
```
Comment (SmallJoker) [inline builtin/game/falling.lua]: Must not be a table. `core.sound_play("sound_name", { pos = p })` for example, is valid too.

File: builtin/game/falling.lua [hunk 2]
```diff
@@ -141,6 +141,11 @@ local function convert_to_falling_node(pos, node)
 	local meta = core.get_meta(pos)
 	local metatable = meta and meta:to_table() or {}
 
+	local def = core.registered_nodes[node.name]
+	if def and def.sounds and def.sounds.fall and def.sounds.fall.name then
```
Comment (SmallJoker) [inline builtin/game/falling.lua]: Same here. Remove the `and def.sounds.fall.name` check.

File: builtin/game/falling.lua (hunk 1 above)
Comment (DS-Minetest) [reply builtin/game/falling.lua]: You wanted to write "Does not need to be a table.", didn't you?
Falling node object becoming a node does the same:
https://github.com/minetest/minetest/blob/766fb7b46edab47f96b8b940e390daf5319261b7/builtin/game/falling.lua#L119-L121

Comment (SmallJoker) [reply builtin/game/falling.lua]: Yes, and yes please.

**Comment by DS-Minetest** [comment]
The comments are handled.

**Comment by paramat** [comment]
Removed my -1. I still disapprove of a sound for an attached node detaching, but a falling node starting to fall is reasonable. Maybe it should fallback to one of the existing sounds at a lower volume 'dig' 'dug'?

**Comment by paramat** [comment]
> It should not be the job of the engine to dictate what kind of artwork is “allowed”.

It is, the engine is inevitably limited.
>  Just because MTG does not need them does not mean that NO game ever would want them.

I'm not writing from a MTG perspective.
> This ideology of “if it doesn't benefit MTG, it is not needed” is really damaging for Minetest.

I didn't express that.

The rest of your arguments are of the type sometimes used which is: 'it's an engine so should do everything and anything', however the engine has to be limited and has to prioritise, especially considering we are open source and therefore very much lacking in core dev time. Simplicity is strongly expressed in the roadmaps.

**Comment by paramat** [comment]
I'm neutral on this now. Sorry for my earlier comment.

**Comment by DS-Minetest** [comment]
No, I think, it won't add
>a noticeable delay between the action of players and the reaction of the server

. https://en.wikipedia.org/wiki/Lag

**Comment by paramat** [comment]
How about naming the sound `node_fall` for clarity, it could be confused with player fall, and might conflict with any player fall sounds added in future.

**Comment by Wuzzy2** [comment]
Good suggestion.

**Comment by DS-Minetest** [comment]
The builtin file is called "falling", not "node_falling", too.
As this is in node def, it seems obvious that the node falls and not a player.
The other sounds are also not called eg. "node_dug".

**Comment by paramat** [comment]
DS-Minetest, sorry you are correct, i must have been confused, it's fine as 'fall'.

**Comment by DS-Minetest** [comment]
So, does someone want to give this PR a 2nd approval?

**Pull request merged**